src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -156,50 +169,59 @@ def _choose_one_version(self, versioned_items: BigQueryRows) -> BigQueryRow:
-                require(associate_id == bundle_project_id,
-                        f'Supplementary file must be associated with the current project '
-                        f'({bundle_project_id}, not {associate_id})')
-                for supp_file in link['files']:
-                    entities['supplementary_file'].add(supp_file['file_id'])
-            else:
-                raise RequirementError(f'Unexpected link_type: {link_type}')
-
-        def retrieve_rows(entity_type: str, entity_ids: Set[str]):
+        def retrieve_links(bundle: TDRBundle, links_id: BundleFQID) -> Tuple[JSON, EntityID]:
+            links_columns = ', '.join(bundle.metadata_columns | {'content', 'project_id', 'links_id'})
+            links_row = one(self._run_sql(f'''
+                SELECT {links_columns}
+                FROM {self._source.bq_name}.links
+                WHERE links_id = '{links_id.uuid}'
+                    AND version = TIMESTAMP('{links_id.version}')
+            '''))
+            links_json = json.loads(links_row['content'])
+            bundle_project_id = links_row['project_id']
+            log.info('Retrieved links content, %s top-level links', len(links_json['links']))
+            bundle.add_entity('links.json', 'links', links_row)
+            return links_json, bundle_project_id
+
+        entities: TypeGroupedEntities = {}
+        unprocessed_fqids = {root_bundle_fqid}
+        processed_fqids = set()
+        # Recursively follow dangling inputs to collect entities from upstream
+        # bundles. Assumes subgraphs connected by dangling edges form a DAG; no
+        # subgraph will be processed more than once, but there is no defense
+        # against cycles.
+        while unprocessed_fqids:
+            # This ought to be concurrent
+            for bundle_fqid in unprocessed_fqids:
+                # Not sure what to do about multiple links.json in the resulting stitched "bundle"
+                links_json, bundle_project_id = retrieve_links(bundle, bundle_fqid)
+                bundle_entities, inputs, outputs = self._scan_links(links_json['links'], bundle_project_id)
+                entities.update(bundle_entities)
+                processed_fqids.add(bundle_fqid)
+                dangling_inputs = self._find_dangling_inputs(inputs, outputs)
+                if dangling_inputs:
+                    log.info('Bundle %r has dangling inputs: %r', bundle_fqid, dangling_inputs)
+                    outputs_to_inputs = {}
+                    for input_type, input_ids in dangling_inputs.items():
+                        outputs_to_inputs.update(self._link_file_ids(input_type, input_ids))
+                    for upstream_bundle_fqid in self._find_upstream_bundles(set(outputs_to_inputs.keys())):
+                        log.info('Upstream bundle %r scheduled for stitching', upstream_bundle_fqid)
+                        unprocessed_fqids.add(upstream_bundle_fqid)
+            unprocessed_fqids -= processed_fqids
```

Comment:
Consider a graph with nodes `root`, `A`, and `B`, with edges `root -> A`, `root -> B`, `A -> B` (where `X -> Y` means "Y is upstream of X"). 
If `unprocessed_fqids = {B, A}` and `B` is processed first, and `B` is removed from `unprocessed_fqids` at the time it is processed, it will be re-added again when `A` is processed and `B` is detected upstream of it, resulting in `B` being processed twice.

Reply:
I wouldn't add B to `unprocessed_fqids` if it's already in `processed_fqids`. In the past, I've written depth-first traversal of graphs with just one set of visited nodes for tracking state. I'm surprised you need two.

Reply:
In a normal traversal, the complete graph is provided as input. So, I'm guessing you did in fact have two data structures: a set of visited nodes `V`, and some other object representing the graph itself, `G = (E, N)`, where `V ⊆ N`.

In this context, we are not provided with `G`, only a root node <code>N<sub>0</sub></code>, and our goal is to incrementally discover `N` by processing the query results. So once again, we have two data structures: `processed_fqids = V`, and `unprocessed_fqids ⊆ N - V` taking the place of `G` as our reservoir of unexplored nodes.